internal/terraform/eval_count.go

```diff
@@ -51,6 +51,18 @@ func evaluateCountExpression(expr hcl.Expression, ctx EvalContext, allowUnknown
 		})
 	}
 
+	// Ephemeral values are not allowed in count expressions.
+	if countVal.HasMark(marks.Ephemeral) {
+		diags = diags.Append(&hcl.Diagnostic{
+			Severity: hcl.DiagError,
+			Summary:  "Invalid count argument",
+			Detail:   "The count value cannot depend on ephemeral values.",
```

Comment:
Nitpick: I would consider avoiding "depend on" in the wording here as that has a specific meaning in Terraform. We could e.g. say `The count value cannot contain ephemeral values.` Relatedly, we should probably be aligning the messages between `for_each` and `count`, so how about this?

```
The given "count" value is derived from an ephemeral value, which means that Terraform cannot persist it between plan/apply rounds. Use only non-ephemeral values here.
```

--- 

On a separate note it would be amazing if in the future we could highlight the part of expression which actually has the mark, rather than the whole expression. I am in no way implying you should be doing that as part of this PR though! 😄

Reply:
I'll reword this 👍